text/0000-custom-panic-handlers.md

```diff
@@ -0,0 +1,125 @@
+
+This is problematic because `panic!` enables returning from a chain of deeply nested functions in an exceptional situation without complicating their signatures.
+
+An example: you have a function, ten levels deep in a child thread, that suddently can encounter a non-recoverable error (after additions to the code). Do you want to refactor ten functions to use `Result` - at a point where you have not yet decided what the best long-term approach is, as it could require heavy refactoring of the existing codebase - or use `panic!` for a while to experiment without hurting user experience (by that I mean, displaying messages best used for debugging to the user - does the user really care about the thread and line number a function failed in?). I'm not advocating the overuse of `panic!` as a cheap error handling mechanism, but it has its uses. Otherwise it would just crash the whole program instead of being possibly caught.
+
+Using `Result`s or `Option`s and creating new enums (to use with `Result`) just to handle these exceptional situations feels overkill, adds a lot of boilerplate while providing little extra functionality, and makes it harder to move around these functions without modifying their return types, while `panic!` gives very similar functionality without all the boilerplate, but since `panic!()` clutters logs with messages appropriate for debugging but not for the end user, one is often forced to resort to the former.
+
+Another concern is that some third-party library functions you have no control upon may panic on exceptional situations, and it is often not desirable to output debugging information to the user in this case.
+
+
+# Detailed design
+
+The current panic handler is defined [here](https://github.com/rust-lang/rust/blob/2b8c9b12f91c0bf2c1e6278a5f803c2df3698432/src/libstd/panicking.rs#L28).
+
+This RFC proposes to allow control over what happens beside unwinding when a thread panic by **changing** the *panic handler* of this thread, instead of simply adding new ones with [`rt::unwind::register`](http://doc.rust-lang.org/std/rt/unwind/fn.register.html).
+
+
+A few common handlers would be added to `std::thread`: **`debug`**, **`basic`** and **`silent`**.  
+The `debug` handler would be the default and have a behavior similar to the current `on_panic`, to avoid modifying the behavior of existing code.
+`basic` would print whatever is passed to `panic!()` without adding information to the message.
+`silent` would never output anything. 
+
+A handler (or a *callback*, as this was called in `rt::unwind`'s code) is registered with `thread::on_panic`. It has the following signature:
+
+``` rust
+fn(msg: &Any + Send, file: &'static str, line: u32) -> bool
```

Comment:
Could you explicitly list the signature of `thread::on_panic` in this RFC as well? Also, it's often quite useful to be able to close over data as well, so it may be nicer to take `F: FnMut(...) + 'static` and stick that in a TLS local.